examples/dataflow-hashpipeline/hashpipeline.py

```diff
@@ -0,0 +1,213 @@
+    parser.add_argument('--secret_name', type=cls.valid_secret,
+      help='Fully qualified name of the Secret Maneger secret where the base64 encoded hash key is stored. i.e. projects/*/secrets/*"')
+    parser.add_argument('--firestore_project',
+      help='Project Id where Firestore database is located and where DLP will be executed.')
+    parser.add_argument('--collection_name',
+      help='Name of the Firestore collection where hashed SSNs are stored')
+    parser.add_argument('--salt',
+      help='Random string for salting hashes.')
+
+  @classmethod
+  def valid_secret(cls, secret):
+    parts = secret.split('/')
+    if len(parts) != 4 or parts[0] != 'projects' or parts[2] != 'secrets':
+      raise ArgumentTypeError("Secret Name must be of format: projects/<PROJECT>/secrets/<SECRET_NAME>")
+    return secret
+
+def run(argv):
+  """Main entry point; defines and runs the pipeline."""
+  opts = HashPipelineOptions()
+  std_opts = opts.view_as(StandardOptions)
+  std_opts.streaming = True 
+
+  def fmt(elem):
+    filename, counts = elem
+    obj = {'filename': filename, 'ssn_counts': counts}
+    return json.dumps(obj).encode('utf-8')
+
+  with beam.Pipeline(options=opts) as p:
+    (p | 'Read filename from Pubsub' >> ReadFromPubSub(None, opts.input_subscription)
+       | 'Decode from UTF-8' >> beam.Map(lambda elem: elem.decode('utf-8'))
+       | 'Read filename and lines from file' >> ReadAllFromTextWithFilename()
+       | 'Get DLP Findings' >> beam.ParDo(DlpFindingDoFn(opts.firestore_project, std_opts.runner))
+       | 'Determine if client SSN' >> beam.ParDo(ExistingSSNsDoFn(
+          opts.firestore_project, opts.salt, opts.secret_name, opts.collection_name, std_opts.runner))
+       | 'Apply windowing' >> beam.WindowInto(FixedWindows(5))
+       | 'Group by Filename' >> beam.GroupByKey()
+       | 'Count SSNs by Filename' >> beam.Map(lambda elem : (elem[0], sum(elem[1])))
+       | 'Format Message' >> beam.Map(fmt)
+       | 'Write to Pubsub' >> WriteToPubSub(opts.output_topic))
+
+class DlpFindingDoFn(beam.DoFn):
+  """Fetch DLP Findings as a PCollection"""
+  def __init__(self, project, runner):
+    beam.DoFn.__init__(self)
+    self.project = project
+    self.runner = runner
+    self.inspect_config = {
+      'info_types'    : [{'name': 'US_SOCIAL_SECURITY_NUMBER'}],
+      'min_likelihood': 'VERY_UNLIKELY',
+      'include_quote' : True # We need the output to match against the KV Store
+    }
+
+  def setup(self):
+    from google.cloud.dlp import DlpServiceClient
+    self.dlp_client = DlpServiceClient()
+
+  def process(self, element):
+    # TODO: Remove when version 2.22.0 is released BEAM-7885
+    if self.runner == 'DirectRunner':
+      self.setup()
+    # Convert the project id into a full resource id.
+    parent = self.dlp_client.project_path(self.project)
+    filename, chunk = element
+    # Call the API.
+    response = self.dlp_client.inspect_content(parent, self.inspect_config, {'value': chunk})
```

Comment:
DLP client uses the default exponential retry. I am wondering if we should only use this if there are too many requests hitting the quota. Else allow it to fail fast.

Reply:
This seems like a good fit:
https://googleapis.dev/python/google-api-core/latest/retry.html#google.api_core.retry.if_transient_error